# Waze app: user churn analysis and prediction model
Part 1: exploratory data analysis (EDA)

In [ ]:
# import packages and libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
# load dataset into a dataframe
df = pd.read_csv("waze_dataset.csv")

In [ ]:
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

The data is already in a structured format. Each row represents a user

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


There are 14,999 rows and 13 columns. 700 rows have label missing. Other variables have no missing values.

Is there a difference between the summary statistics of the 700 rows with missing labels and the rows that are not missing any values?


In [ ]:
null_df = df[df["label"].isnull()]

null_df.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [ ]:
not_null_df = df[-df["label"].isnull()]

not_null_df.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

No  discernible differences are found: the means and standard deviations are fairly consistent between the two groups.

Next, we may want to check how many iPhone users vs Android users have null values.

In [ ]:
null_df["device"].value_counts()

iPhone     447
Android    253
Name: device, dtype: int64

In [ ]:
null_df["device"].value_counts(normalize=True)

iPhone     0.638571
Android    0.361429
Name: device, dtype: float64

On a total of 700 rows with null values, 447 are iPhone users (64%) and 253 are Android users (36%).

In [ ]:
# calculate the percentage of iphone vs android users in the whole dataset
df["device"].value_counts(normalize=True)

iPhone     0.644843
Android    0.355157
Name: device, dtype: float64

By looking at the whole dataset, we find that the percentage of iPhone users vs Android users is consistent with the percentage found when analyzing only the missing values for each device. Therefore, we do not see any obvious pattern to the missing values.

Next, we look at the counts and percentages of users who churned vs those who were retained.

In [ ]:
df["label"].value_counts()

retained    11763
churned      2536
Name: label, dtype: int64

In [ ]:
df["label"].value_counts(normalize=True)

retained    0.822645
churned     0.177355
Name: label, dtype: float64

The dataset contains 11763 (82%) retained users and 2536 (18%) churned users.

Next, we calculate the median values by label (calculating the mean would be subject to the influence of outliers, while the median represents the middle value of the distribution regardless of any outlying values).

In [ ]:
df.groupby("label").median(numeric_only=True)

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

By comparing the medians of each variable for churned and retained users, we find that:

*   churned users performed an average of 3 more drives in the last month compared to the number of drives of retained users.
*   retained users used the app more than twice as many days as churned users.
*   the median churned user drove ~200 more km and for ~2.5 more hours than the median retained user.
*   churned users had more drives in fewer days (an average of 50 drives in 6 days) compared to retained users (with an average of 47 drives in 16 days).

This suggests that churned users had more and longer drives compared than the average retained user.


Based on the above considerations, we want to find for both groups:


*   the median kilometers per drive
*   the median kilometers per driving day
*   the median drives per driving day




In [ ]:
# calculate the median km per drive in the last month for both retained and churned users
median_by_label = df.groupby("label").median(numeric_only=True)

print('Median kilometers per drive:')
median_by_label["driven_km_drives"] / median_by_label["drives"]

Median kilometers per drive:


label
churned     73.053113
retained    73.716694
dtype: float64

We can see that the median user, whether churned or retained, drove an average of 73 km/drive.

In [ ]:
print('Median kilometers per driving day:')
median_by_label["driven_km_drives"] / median_by_label["driving_days"]

Median kilometers per driving day:


label
churned     608.775944
retained    247.477472
dtype: float64

In [ ]:
print('Median drives per driving day:')
median_by_label["drives"] / median_by_label["driving_days"]

Median drives per driving day:


label
churned     8.333333
retained    3.357143
dtype: float64

We see a significant difference in the kilometers per driving day of the median churned user (608km), compared to the kilometers per day of the median retained user (247km). Similarly, the median churned user had a significantly higher number of drives per day compared to the median retained user.

In [ ]:
# calculate how many users churned by device type
df.groupby(["label", "device"]).size()

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

In [ ]:
# calculate the percentage of how many users churned by device type
df.groupby("label")["device"].value_counts(normalize=True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: device, dtype: float64

As we can see, the ratio of iPhone users and Android users is consistent between the churned group and the retained group. Therefore, nothing suggests that churn is correlated to the device used.
